### ISSUE TITLE
Feature request: RESP3 protocol support

### ISSUE THREAD

**Issue by matobet** [informatikr/hedis#156]
Are there any plans to support the new [RESP3 protocol](https://github.com/antirez/RESP3/blob/master/spec.md) introduced in Redis 6?

Supporting it (along with its PUSH messages) would be helpful in supporting the [Server-assisted client side caching](https://redis.io/topics/client-side-caching) in the future.

**Comment by k-bx** [comment]
@matobet would you care enough to work on it? ;)

**Comment by matobet** [comment]
I suppose with the proper guidance I could give it a shot 😅

**Comment by k-bx** [comment]
Sure, feel free to ask for help or a review, happy to assist!

**Comment by aravindgopall** [comment]
Hi @matobet, have you got any chance to work on this?

**Comment by matobet** [comment]
@aravindgopall unfortunately no :-(

**Comment by 414owen** [comment]
I wrote a RESP3 parser here: https://github.com/414owen/resp

RESP3 adds quite a lot of new syntactic forms, but it's still possible to parse it without backtracking.

It should be complete, but tomorrow I'll double-check the spec, write a few more property tests, and upload it to hackage.

**Comment by 414owen** [comment]
Yeah the pub/sub listener needs to inspect messages constantly, to know if they're a push message. This means we don't get pipelining.

One way around that is to have a connection type specifically for pub.sub with commands. This would amount to one pipelined connection, and one not. It could be fairly transparent to the user, which is nice. 

I think that splitting the connection up (again, there's already the clustered one), really necessitates fixing [this issue](https://github.com/informatikr/hedis/issues/219) and parameterizing operations over the connection types they can run on.
